docs: OpenID Connect under federated login. (#2376)
mendersoftware/mender-docs: 08.Server-integration/09.OIDC-Federated-Authentication/docs.md

---
title: OpenID Connect Federated Authentication
taxonomy:
    category: docs
---

!!!!! OpenID Connect Federated Authentication is only available in the Mender Enterprise plan.
!!!!! See [the Mender features page](https://mender.io/product/features?target=_blank)
!!!!! for an overview of all Mender plans and features.

> OpenID Connect (OIDC) uses a very similar flow to [SAML](../08.SAML-Federated-Authentication).
> And as with SAML, you can configure Mender to initiate the authorisation process
> with the OpenID Connect provider by exposing a login endpoint which redirects
> user to the authorisation server which in turn calls authentication callback.

Reviewer: ```suggestion
OpenID Connect (OIDC) uses a very similar flow to [SAML](../08.SAML-Federated-Authentication).
And as with SAML, you can configure Mender to initiate the authorization process
with the OpenID Connect provider by exposing a login endpoint that redirects
user to the authorization server, which in turn calls an authentication callback.
```

Author: fixed.

> Please note that currently Mender implements [Implicit Flow](https://openid.net/specs/openid-connect-implicit-1_0.html).

Reviewer: ```suggestion
Please note that as of today, Mender only implements the [Implicit Flow](https://openid.net/specs/openid-connect-implicit-1_0.html).
```

Author: fixed.

## Configure OpenID Connect: create a provider

The first step is to create a _provider_. Please note that there are many names for actors
in all the federated authentication mechanisms. Mender mostly follows SAML nomenclature.
The provider we are going to create is an entity that stores all the settings needed
to initiate and complete the OIDC flow.

### The well-known endpoint

> There is a number of parameters needed for OIDC to operate. You can get all of them from the so-called
> _well-known_ endpoint, and use those to create a provider, or pass the well-known url in the creation
> process. In the latter case you need to use the following document (either
> in the [organization-and-billing](https://hosted.mender.io/ui/settings/organization-and-billing)
> or in the POST API call to [https://hosted.mender.io/api/management/v1/useradm/sso/idp/metadata](https://docs.mender.io/api/#management-api-user-administration-and-authentication-post-saml--openid-connect-metadata--or-url-point-to-them)
> endpoint)

Reviewer: ```suggestion
There are several parameters needed for OIDC to operate. You can get all of them from the so-called
_well-known_ endpoint, and use those to create a provider or pass the well-known URL in the creation
process. In the latter case, you need to use the following document (either
in the [organization-and-billing](https://hosted.mender.io/ui/settings/organization-and-billing)
or in the POST API call to [https://hosted.mender.io/api/management/v1/useradm/sso/idp/metadata](https://docs.mender.io/api/#management-api-user-administration-and-authentication-post-saml--openid-connect-metadata--or-url-point-to-them)
endpoint)
```

Author: fixed.

> ### Upload the metadata and get the login url

Reviewer: ```suggestion
### Upload the metadata and get the login URL
```

Author: fixed.

> In order to create an OIDC provider in Mender you can use the UI
> [organization-and-billing](https://hosted.mender.io/ui/settings/organization-and-billing).
> Enabling the Single Sign-On and choosing the type will bring you the possibility to type or upload
> the metadata you need to configure OpenID Connect as a way to authenticate users.

Reviewer: ```suggestion
To create an OIDC provider in Mender, you can use the UI accessing the [Organization and billing](https://hosted.mender.io/ui/settings/organization-and-billing) view.
Enabling the Single Sign-On and choosing the type will allow you to type or upload the metadata you need to configure OpenID Connect to authenticate users.
```

Author: fixed.

> Mender can use the well-known url, and in that case it is enough to provide the following
> document:

Reviewer: ```suggestion
Mender can use the well-known URL, and in that case, it is enough to provide the following
document:
```

Author: fixed.

```json
{
  "name":"identity-provider-name",
  "client_id":"Ylb6gfnmXckxHFoXH1aKLX2poDCvS9MV",
  "client_secret":"zygGHMTL9VlCQpOIHNbXpTZqd77LqqIP",
  "well_known_url":"https://your.openidconnect.provider.oidc/.well-known/openid-configuration"
}
```

> In the above the name is obligatory while both `client_id` and `client_secret` are optional, but maybe required
> by your OpenID Connect provider. The `well_known_url` denotes the URL for the well-known endpoint which contains
> the hostname of your OIDC provider (called "your.openidconnect.provider.oidc" throughout this section).

Reviewer: ```suggestion
In the above, the name is mandatory, while both `client_id` and `client_secret` are optional but may be required by your OpenID Connect provider. The `well_known_url` denotes the URL for the well-known endpoint, which contains the hostname of your OIDC provider (called "your.openidconnect.provider.oidc" throughout this section).
```

Author: fixed.

      "code",
      "token",
      "id_token",
      "code token",
      "code id_token",
      "token id_token",
      "code token id_token"
    ],
    "code_challenge_methods_supported": [
      "S256",
      "plain"
    ],
    "response_modes_supported": [
      "query",
      "fragment",
      "form_post"
    ],
    "subject_types_supported": [
      "public"
    ],
    "id_token_signing_alg_values_supported": [
      "HS256",
      "RS256",
      "PS256"
    ],
    "token_endpoint_auth_methods_supported": [
      "client_secret_basic",
      "client_secret_post",
      "private_key_jwt"
    ],
    "claims_supported": [
      "aud",
      "auth_time",
      "created_at",
      "email",
      "email_verified",
      "exp",
      "family_name",
      "given_name",
      "iat",
      "identities",
      "iss",
      "name",
      "nickname",
      "phone_number",
      "picture",
      "sub"
    ],
    "request_uri_parameter_supported": false,
    "request_parameter_supported": false,
    "token_endpoint_auth_signing_alg_values_supported": [
      "RS256"
    ],
    "end_session_endpoint": "https://your.openidconnect.provider.oidc/oidc/logout"
  }
}

> }

Reviewer: Extra closed parenthesis

```suggestion
```

Author: fixed. nicely caught!

```

> In the former case, Mender will query and validate the required settings.
> Once you see the success on parsing and saving the metadata the UI will
> present you with a start or login url: this is the address you use
> to initiate the login process and this completes the setup.

Reviewer: ```suggestion
In the first case, Mender will query and validate the required settings.
Once you see the success of parsing and saving the metadata, the UI will
present you with a start or login URL: this is the address you use
to initiate the login process, which completes the setup.
```

Author: fixed.
